invoices/invoiceregistry_test.go

```diff
@@ -654,18 +542,77 @@ func TestUnknownInvoice(t *testing.T) {
 	hodlChan := make(chan interface{})
 	amt := lnwire.MilliSatoshi(100000)
 	_, err := ctx.registry.NotifyExitHopHtlc(
-		hash, amt, testHtlcExpiry, testCurrentHeight,
+		testInvoicePaymentHash, amt, testHtlcExpiry, testCurrentHeight,
 		getCircuitKey(0), hodlChan, testPayload,
 	)
 	if err != channeldb.ErrInvoiceNotFound {
 		t.Fatal("expected invoice not found error")
 	}
 }
 
-type mockPayload struct {
-	mpp *record.MPP
-}
+// Tests that invoices are canceled after expiration.
+func TestInvoiceExpiryWithRegistry(t *testing.T) {
+	t.Parallel()
 
-func (p *mockPayload) MultiPath() *record.MPP {
-	return p.mpp
+	ctx := newTestContext(t)
+	defer ctx.cleanup()
+
+	// Subscribe to the not yet existing invoice.
+	subscription, err := ctx.registry.SubscribeSingleInvoice(
+		testInvoicePaymentHash,
+	)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer subscription.Cancel()
+
+	if subscription.hash != testInvoicePaymentHash {
+		t.Fatalf("expected subscription for provided hash")
+	}
+
+	testExpiry := time.Hour
+
+	// Create the invoice that has already been expired.
+	timestamp := time.Date(2017, time.July, 21, 10, 0, 0, 0, time.UTC)
+	expiringInvoice := newTestInvoice(t, timestamp, testExpiry)
+
+	// Add the invoice and expect that AddInvoice is successful.
+	addIdx, err := ctx.registry.AddInvoice(
+		expiringInvoice, testInvoicePaymentHash,
+	)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if addIdx != 1 {
+		t.Fatalf("expected addIndex to start with 1, but got %v",
+			addIdx)
+	}
+
+	// First we expect the open state to be sent to the single invoice subscriber.
+	select {
+	case update := <-subscription.Updates:
+		if update.State != channeldb.ContractOpen {
+			t.Fatalf("expected state ContractOpen, but got %v",
+				update.State)
+		}
+	case <-time.After(testTimeout):
+		t.Fatal("no update received")
+	}
```

Comment:
May I shill `defer timeout(t)()` to you as well? I personally don't like the bloat of selecting on timeout channels.

Reply:
This test is now updated. I'd love to use `defer timeout(t)()` but we need to actually wait here before the expectation checks.